rt_waterfall_D3S.py

```diff
@@ -0,0 +1,112 @@
+from auxiliaries import set_verbosity
+import time
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+class Rt_Waterfall_D3S(object):
+    """
+    Class for running the D3S in real-time waterfall mode
+    """
+    
+    def __init__(self, 
+                 manager=None, 
+                 verbosity=1,
+                 logfile=None,
+                ):
+        
+        self.v = verbosity
+        if manager and logfile is None:
+            set_verbosity(self, logfile=manager.logfile)
+        else:
+            set_verbosity(self, logfile=logfile)
+            
+        self.manager = manager
+        
+        self.interval = manager.interval
+        
+        self.queuelength = None
+        self.image = None
+    
+    def get_data(self, spectra):
+        '''
+        Transfers spectra data to 2 queues that were designed for 
+        real time waterfall mode.
+        Call rebin spectra in this method.
+        '''
+        new_spectra = self.rebin(spectra)
+        self.manager.wqueue2.append(new_spectra)
+        self.queuelength = len(self.manager.wqueue2)
+        for i in self.manager.wqueue2:
+            self.manager.wqueue1.append(i)
+   
+    def rebin(self, data, n=4):
+        """
+        Rebins the array. n is the divisor. Rebin the data in the grab_data method.
+        """
+        a = len(data)/n
+        new_data = np.zeros((256, 1))
+        i = 0
+        count = 0
+        while i < a:
+            temp = sum(data[i:n*(count+1)])
+            new_data[count] = temp
+            count += 1
+            i += n
+        return new_data
+
+    def fix_array(self, array):
+        """
+        Used to format arrays for the waterfall plot.
+        Called inside make_image.
+        """
+        new_array = np.zeros((256))
+        i = 0
+        while i < 256:
+            new_array[i] = array[i]
+            i += 1
+        return new_array
+      
+    def make_image(self):
+        """
+        Prepares an array for the waterfall plot
+        Call fix_array in this method
+        """
+        self.image = np.zeros((self.queuelength, 256),dtype=float)
+        j = 0
+        while j < self.queuelength:
+            i = 0
+            temp = self.fix_array(self.manager.wqueue1.pop())
+            while i < 256:
+                self.image[j][i] = temp[i]
+                i += 1
```

Comment:
You should be able to do this with numpy slices which will be faster. If the dimensions don't match, use `flatten` or `squeeze` or so. Something like `self.image[j, :] = temp`.